File: src/qlearn.py
import numpy as np  # pylint: disable=import-error
import random
from gym import utils  # pylint: disable=import-error
from io import StringIO
import sys
from contextlib import closing
import copy

import gym  # pylint: disable=import-error
import time
import random
from collections import deque


from taxienv import TaxiEnv

map = [
    "+---------+",
    "|R: | : :G|",
    "| : | : : |",
    "| : : : : |",
    "| | : | : |",
    "|Y| : |B: |",
    "+---------+",
]


class QAgent():
    def __init__(self, env):
        self.env = env
        self.alpha = 0.25
        self.epsilon = 1
        self.rate = 1
        self.q = {}


    def actions(self, state):
        s = self.env.decode(state)
        break_down = (s[0], s[1])
        if break_down in self.env.special:
            ls = [i for i in range(10)]
            return ls
        else:
            ls = [i for i in range(6)]
            return ls


    def get_q_value(self, state, action):
        if action not in self.actions(state):
            raise ValueError

        if state not in self.q:
            return 0

        elif action not in self.q[state]:
            return 0

        else:
            return self.q[state][action]


    def best_reward_state(self, state):
        opt = float("-inf")
        ls = self.actions(state)

        if state not in self.q:
            return 0

        for a in ls:
            if a in self.q[state]:
                if self.q[state][a] > opt:
                    opt = self.q[state][a]
            else:
                if opt < 0:  # if (state, a) not in self.q, we default a q-value of 0
                    opt = 0

        return opt


    def choose_action(self, state, prob = True):
        values = [0, 1]
        probs = [self.epsilon, 1 - self.epsilon]
        rand = random.choices(values, probs)
        ls = self.actions(state)

        if not prob or rand == 1:
            opt = float("-inf")
            action = None
            for a in ls:
                value = self.get_q_value(state, a)
                if value > opt:
                    opt = value
                    action = a
            return action

        else:
            action = random.choice(ls)
            return action


    def qlearn(self, num_episodes, show=True, number=None, render=True):

        for i in range(num_episodes):
            if show and (i % 100 == 0 or i == num_episodes - 1):
                if number is None:
                    print("Episode {} begins!".format(i))
                else:
                    print("Episode {} begins! ({})".format(i, number))

            s = self.env.reset()
            t = 0

            while True and t < 800:
                action = self.choose_action(s)
                s_next, reward, done = self.env.step(action)

                future_rewards_estimated = self.best_reward_state(s_next)
                old_q = self.get_q_value(s, action)

                # Update q-value
                if s not in self.q:
                    self.q[s] = {}

                self.q[s][action] = old_q + self.alpha * (reward + self.rate * future_rewards_estimated - old_q)

                s = s_next
                if done:
                    break

                t += 1

            if show and (i % 100 == 0 or i == num_episodes - 1):
                if number is None:
                    print("Episode {} done ({})!".format(i, t))
                else:
                    print("Episode {} done ({})({})!".format(i, t, number))
                
                if render:
                    self.env.render()

            self.epsilon *= 0.995


    def eval(self, show=True, fixed=None):
        s = self.env.reset()
        if fixed is not None:
            self.env.current = fixed
            s = self.env.current
        if show:
            self.env.render()
        steps = []
        bd = self.env.decode(s)
        states = [(bd[0], bd[1])]
        t = 0
        total = 0

        while t < 1000:
            action = self.choose_action(s, prob = False)
            s_next, reward, done = self.env.step(action)
            steps.append(action)
            break_down = self.env.decode(s_next)
            states.append((break_down[0], break_down[1]))
            total += reward

            s = s_next
            t += 1

            if done:
                break

        if show:
            self.env.render()
        return [steps, total, states]


    def print_eval_result(self, output):
        # Print results from evaluation for visualization
        print("Steps taken: {}".format(output[0]))
        print("Total reward: {}".format(output[1]))
        print("States traversed: {}".format(output[2]))

        return


'''map_to_numpy = np.asarray(map, dtype='c')
env = TaxiEnv(map_to_numpy)

modified = env.transition([(1, 4), (5, 2), (4, 2)])
modified.special.append((1, 2))
modified.special.append((2, 4))


agent = QAgent(modified)
start = time.time()
agent.qlearn(700)
end = time.time()

series = env.resettable_states()
vals = []

for state in series:
    res = agent.eval(fixed=state, show=True)
    vals.append(res[1])

# Agent with less training
new_agent = QAgent(modified)
start = time.time()
new_agent.qlearn(600, show=False, render=False)
end = time.time()

series = env.resettable_states()
n_vals = []

for state in series:
    res = new_agent.eval(fixed=state, show=False)
    n_vals.append(res[1])

a = sum(n_vals) - sum(vals)

print(a)'''

File: c-conversion/qlearn.cpp
#include <iostream>
#include <vector>
#include <time.h>
#include <random>
#include <algorithm>
#include <bits/stdc++.h>
#include <numeric>
#include <utility>
#include <unistd.h>
#include <cstdlib>

#include "wenv.h"
using namespace std;

int choice(vector <int> v, vector <float> probs)
{
    assert(v.size() == probs.size());

    // Random number
    float val = (float)rand() / RAND_MAX;

    // Partial sums (cumulative sums)
    float result[probs.size()];

    partial_sum(probs.begin(), probs.end(), result);

    int a = -1;

    if (val < result[0]) a = v[0];

    else  // should have been binary search, but not due to developer's laziness
    {
        for (int i = 0; i < probs.size() - 1; i++)
        {
            if (result[i] <= val && val < result[i + 1])
            {
                a = v[i + 1];
                break;
            }
        }
    }

    return a;
}

struct w_QAgent
{
    WindyGridworld env;
    float alpha = 0.5;
    float epsilon = 1;
    float rate = 1.0;

    map <vector<int>, map <int, float>> q;

    w_QAgent(WindyGridworld e)
    {
        env = e;
    }

    vector <int> actions(vector <int> state)
    {
        vector <int> pos{state[0], state[1]};
        bool special_cell = false;
        bool jump_cell = false;

        int a = env.special.size();
        int b = env.jump_cells.size();

        if (a != 0)
        {
            for (int ind = 0; ind < a; ind++)
            {
                if (pos == env.special[ind])
                {
                    special_cell = true;
                    break;
                }
            }
        }

        if (b != 0)
        {
            for (int ind = 0; ind < b; ind++)
            {
                if (pos == env.jump_cells[ind])
                {
                    jump_cell = true;
                    break;
                }
            }
        }

        vector <int> actions;

        if (special_cell && jump_cell)
        {
            for (int ind = 0; ind < 12; ind++)
            {
                actions.push_back(ind);
            }
        }

        else if (special_cell)
        {
            for (int ind = 0; ind < 8; ind++)
            {
                actions.push_back(ind);
            }
        }

        else if (jump_cell)
        {
            for (int ind = 0; ind < 4; ind++)
            {
                actions.push_back(ind);
            }

            for (int ind = 8; ind < 12; ind++)
            {
                actions.push_back(ind);
            }
        }

        else
        {
            for (int ind = 0; ind < 4; ind++)
            {
                actions.push_back(ind);
            }
        }

        return actions;
    }

    float get_q_value(vector <int> state, int action)
    {
        map <vector<int>, map <int, float>>::iterator it;
        it = q.find(state);

        if (it == q.end()) return 0.0;

        else
        {
            map <int, float>::iterator it_1;
            it_1 = q[state].find(action);

            if (it_1 == q[state].end()) return 0.0;

            else return q[state][action];
        }
    }

    float best_reward_state(vector <int> state)
    {
        map <vector<int>, map <int, float>>::iterator it;
        it = q.find(state);

        if (it == q.end()) return 0.0;

        else
        {
            float opt_val = q[state].begin()->second;

            for (auto it_1 = q[state].begin(); it_1 != q[state].end(); it_1++)
            {
                float x = it_1->second;
                if (x > opt_val)
                {
                    opt_val = x;
                }
            }

            if (q[state].size() == actions(state).size()) return opt_val;

            else return max(opt_val, (float)0);
        }
    }

    int choose_action(vector <int> state, bool prob=true)
    {
        vector <int> vals{0, 1};
        vector <float> probs{epsilon, 1 - epsilon};

        int r = choice(vals, probs);
        vector <int> ls = actions(state);

        if (!prob || r == 1)
        {
            float opt_val = (float)INT_MIN;
            int action = -1;
            int N = ls.size();

            for (int i = 0; i < N; i++)
            {
                float value = get_q_value(state, ls[i]);
                if (value > opt_val)
                {
                    opt_val = value;
                    action = ls[i];
                }
            }

            assert(action != -1);
            return action;
        }

        else
        {
            int ind = randint(0, ls.size() - 1);
            return ls[ind];
        }
    }

    void qlearn(int num_episodes, bool show=true, int number=-1, bool render=true)
    {
        for (int i = 0; i < num_episodes; i++)
        {
            // srand(i);
            if (show && (i % 100 == 0 || i == num_episodes - 1))
            {
                if (number == -1) printf("Episode %d begins!\n", i);

                else printf("Episode %d begins! (%d)\n", i, number);
            }

            vector <int> s = env.reset();
            int t = 0;

            while (t < 2000)
            {
                if (t % 100 == 0) srand(2000 * i + t);
                int action = choose_action(s);
                holder h = env.step(action);

                vector <int> s_next = h.s;
                float reward = h.r;
                bool done = h.d;

                float future_rewards_estimated = best_reward_state(s_next);
                float old_q = get_q_value(s, action);

                // Update q-value
                q[s][action] = old_q + alpha * (reward + rate * future_rewards_estimated - old_q);

                s = s_next;

                if (done) break;

                t += 1;
            }

            if (show && (i % 100 == 0 || i == num_episodes - 1))
            {
                if (number == -1) printf("Episode %d done! (%d)\n", i, t);

                else printf("Episode %d begins! (%d) (%d)\n", i, t, number);

                if (render) env.render();
            }
        }

        if (epsilon > 0.05) epsilon *= 0.995;
    }

    pair <float, vector <vector <int>>> eval(bool display, vector <int> fixed)
    {
        vector <int> s = env.reset();

        // Use fixed vector
        env.current = fixed;
        s = env.current;

        vector <vector <int>> states{s};

        int t = 0;
        float total = 0.0;

        while (t < 1000)
        {
            int action = choose_action(s, false);
            holder h = env.step(action);

            vector <int> s_next = h.s;
            float reward = h.r;
            bool done = h.d;

            states.push_back(s_next);
            total += reward;

            s = s_next;
            t += 1;

            if (done) break;
        }

        if (display) env.render();

        return make_pair(total, states);
    }
};

int main()
{
    WindyGridworld env;

    // Modify environment
    vector <int> p{1, 5};
    env.jump_cells.push_back(p);
    w_QAgent agent(env);

    clock_t start = clock();
    agent.qlearn(3000, true, -1, false);
    clock_t end = clock();

    vector <vector <int>> ls = agent.env.resettable_states();
    vector <float> values;

    for (int i = 0; i < ls.size(); i++)
    {
        float x = agent.eval(false, ls[i]).first;
        values.push_back(x);
    }

    cout << "Time taken: " << (float)(end - start) / CLOCKS_PER_SEC << endl;

    float sum = 0.0;
    for (int i = 0; i < values.size(); i++)
    {
        sum += values[i];
    }

    cout << (float)sum / values.size() << endl;
    agent.env.render();

    return 0;
}